### DISCUSSION TITLE
Reusable Model Code Block

### DISCUSSION THREAD

**Discussion by society-research** [FLAMEGPU/FLAMEGPU2#1214] [Ideas]
Hi,

in the spirit of [1] do you guys think there's a good way towards unified code reuse for FLAMEGPU2 models?

Background:
After reading https://github.com/FLAMEGPU/FLAMEGPU2/discussions/797 , [it seems](https://github.com/FLAMEGPU/FLAMEGPU2/discussions/797#discussioncomment-2159552) that FLAMEGPU2 cannot solve issues that are related to jitify's include handling for large include hierarchies. But could a modular code-reuse schema with minimal includes be implemented in cmake? (Possibly collapsing reusable code into a single include file?) I'm just hoping for a unified way forward for larger future research projects. So I just wanted to ask / start the dicussion. It's dawning on me that there's not much hope until https://github.com/NVIDIA/jitify/pull/131 is done.

Kind regards,
Laurin 

[1] https://www.sciencedirect.com/science/article/pii/S1364815224000641

**Comment by mondus** [comment]
Hi Laurin. Thanks for you question and use of FLAME GPU. 

I am not exactly sure what you mean by " a modular code-reuse schema with minimal includes be implemented in cmake". It is already possible to specify additional include directories for JITIFY. See `FLAMEGPU_RTC_INCLUDE_DIRS` option in the Readme.. The obvious limitation to this is that anything included would have to be "safe" for device compilation. E.g. You couldn't just include arbitrary code and expect it to run within a GPU kernel but you could write your own re-usable bits of device code for agent function to use within different models.

**Reply by society-research** [reply]
Sorry, my formulation was not very clear!

More background:
CMake is a build system, that also handles dependencies to other pieces of source code or compiled libraries.
But CMake is chaos. If two people sit in front of a task to do a certain thing in CMake, then the results will look vastly different.

Let's say I want to use some CUDA code from some repository in github, I add the repo to CMake to be downloaded, then in FLAMEGPU2 I can set `FLAMEGPU_RTC_INCLUDE_DIRS` and extend it for every repo I add. Then I recompile. That's already a bit strange that FLAMEGPU is recompiled. But ignoring that there is no unified structure for a github repo that only provides agent function CUDA. So for another github repo I cannot reuse the CMake code I wrote for the first repo.

I was just wondering if you have ever thought about creating some default directory structure for agent function code & includes. Good would be to have said structure with the option to extend it with external code (which of course is all CUDA agent function code).
My issue is just the same as in the discussions that I mentioned earlier, that appart from the missing CMake structure of a model project, there's also the issue of compile time for large include hierarchies.

If you have more question, maybe we can have a call, that might be easier?

**Reply by Robadob** [reply]
>  I add the repo to CMake to be downloaded, then in FLAMEGPU2 I can set FLAMEGPU_RTC_INCLUDE_DIRS and extend it for every repo I add. Then I recompile. That's already a bit strange that FLAMEGPU is recompiled.

`FLAMEGPU_RTC_INCLUDE_DIRS` is an operating system environment variable, it has nothing to do with CMake.

https://github.com/FLAMEGPU/FLAMEGPU2/#environment-variables

If not specified, the working directory at runtime will be used by default.

**Reply by mondus** [reply]
Drop me an email on p.richmond@sheffield.ac.uk and we can arrange a call to discuss.

**Reply by society-research** [reply]
> >  I add the repo to CMake to be downloaded, then in FLAMEGPU2 I can set FLAMEGPU_RTC_INCLUDE_DIRS and extend it for every repo I add. Then I recompile. That's already a bit strange that FLAMEGPU is recompiled.
> 
> `FLAMEGPU_RTC_INCLUDE_DIRS` is an operating system environment variable, it has nothing to do with CMake.
> 
> https://github.com/FLAMEGPU/FLAMEGPU2/#environment-variables
> 
> If not specified, the working directory at runtime will be used by default.

Oh my bad, thanks for correcting! 🙇‍♂️
